HotKey/HotKey/KeyBoardHook.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+
+namespace HotKey
+{
+    public sealed class KeyboardHook : IDisposable
+    {
+        /*
+         * Quick rundown of the Dllimport code here. Basically DLL is an extenal class
+         * that has a ton of helpful methods (Register and UnregisterHotkey for example)
+         * So we call it and pass in the paramters. If you wanted to see the code you can
+         * google Dll registerhotkey code online, but its actually nonsense so don't bother
+         */
+
+        // Registers a hot key with Windows.
+        [DllImport("user32.dll")]
+        private static extern bool RegisterHotKey(IntPtr hWnd, int id, uint fsModifiers, uint vk);
+        // Unregisters the hot key with Windows.
+        [DllImport("user32.dll")]
+        private static extern bool UnregisterHotKey(IntPtr hWnd, int id);
+
+        //Error hanndler for any hot keys we do not have
+        internal void RegisterHotKey(Keys keys, Keys keys2)
+        {
+            throw new NotImplementedException();
+        }
+        //this is the window that we want to actually receive the message (aka our app)
+        private class Window : NativeWindow, IDisposable
+        {
+            //required line of code from mdsn link here
+            //https://msdn.microsoft.com/en-us/library/windows/desktop/ms646279(v=vs.85).aspx
+            private static int WM_HOTKEY = 0x0312;
+
+            public Window()
+            {
+                // create the handle for the window.
+                this.CreateHandle(new CreateParams());
+            }
+
+            /// <summary>
+            /// Overridden to get the notifications.
+            /// </summary>
+            /// <param name="m"></param>
+            protected override void WndProc(ref Message m)
+            {
+                base.WndProc(ref m);
+
+                // check if we got a hot key pressed.
+                if (m.Msg == WM_HOTKEY)
+                {
+                    // get the keys.
+                    Keys key = (Keys)(((int)m.LParam >> 16) & 0xFFFF);
+                    ModifierKeys modifier = (ModifierKeys)((int)m.LParam & 0xFFFF);
+
+                    // invoke the event to notify the parent.
+                    if (KeyPressed != null)
+                        KeyPressed(this, new KeyPressedEventArgs(modifier, key));
+                }
+            }
+
+            public event EventHandler<KeyPressedEventArgs> KeyPressed;
+
+            #region IDisposable Members
+
+            public void Dispose()
+            {
+                this.DestroyHandle();
+            }
+
+            #endregion
+        }
+
+        private Window _window = new Window();
```

Comment:
move these two fields to top of class for readability?

Reply:
I'm getting weird errors when I try to move it around. I'll show you later